source/renderer/app/components/widgets/forms/PinCode.js

```diff
@@ -31,42 +39,258 @@ export default class PinCode extends Component<Props> {
+  onChange = (inputValue: ?number | ?string, key: number) => {
     const { value, onChange } = this.props;
+    const { isBackSpace } = this.state;
     const inputNewValue =
-      inputValue !== null && inputValue !== undefined
+      inputValue !== null && inputValue !== undefined && !isNaN(inputValue)
         ? inputValue.toString()
         : '';
-
-    if (
-      !Object.prototype.hasOwnProperty.call(value, key) ||
-      value[key] === '' ||
-      inputNewValue === ''
-    ) {
+    if (this.hasInputNewValue(inputNewValue, key)) {
       const newValue = value;
-      newValue[key] = inputNewValue;
+      if (!isNaN(inputValue)) {
+        // Recheck if user pressed backspace and moved cursor to previous input field which has value
+        if (
+          isBackSpace &&
+          newValue[key] !== '' &&
+          inputNewValue === '' &&
+          this.focusKey !== key
+        ) {
+          // Set old value to field because it was previously deleted while shifting focus
+          newValue[key] = value[key];
+          // Calculate new field key for focus
+          const focusKey =
+            this.inputsRef[key] &&
+            this.inputsRef[key].inputElement.current.selectionStart === 0 &&
+            key > 2 &&
+            key < this.inputsRef.length
+              ? key - 1
+              : key;
+          // Delay focus to the same field while waiting for the validation to pass
+          setTimeout(() => {
+            const inputFieldRef = this.inputsRef[focusKey];
+            if (inputFieldRef && inputFieldRef.inputElement) {
+              this.setFocusOnField(inputFieldRef);
+              this.setState({ focusKeyChanged: false, focusIsUpdated: true });
+            }
+          }, 0);
+        } else {
+          // Set new value to input field when focus was not shifted to previous field
+          newValue[key] = inputNewValue;
+          this.setState({
+            isBackSpace: false,
+            focusKeyChanged: false,
+            focusIsUpdated: false,
+          });
+        }
+      }
       if (onChange) {
-        onChange(newValue);
+        // Send new updated value to onChange event
+        onChange(newValue, inputNewValue, key);
       }
+      // Update focus key
       this.focusKey = key;
-      this.add = inputValue !== null && inputValue !== undefined;
+      // Recheck if user is adding or deleting value
+      this.add =
+        inputValue !== null && inputValue !== undefined && !isNaN(inputValue);
     }
   };
 
   componentDidUpdate() {
-    const { value, length } = this.props;
+    const { value, length, name, selectedPinField } = this.props;
+    const { isBackSpace, focusKeyChanged, focusIsUpdated } = this.state;
     const key = value.join('').length;
-    if (key > 0 && key < length) {
-      const inputFocusKey = this.add ? this.focusKey + 1 : this.focusKey - 1;
+    const inputValue = value[key - 1];
+    this.add = this.fromBackspace
```

Comment:
Can we change this `this.add` to something more descriptive. What we are adding?

Reply:
Changed it to this.isAddingNewValue